### PR TITLE
Better attribute and metaitem encapsulation throughout the compiler

### PR DESCRIPTION
**Pull request by cgswords**
This PR refactors most (hopefully all?) of the `MetaItem` interactions outside of `libsyntax` (and a few inside) to interact with MetaItems through the provided traits instead of directly creating / destruct / matching against them. This is a necessary first step to eventually converting `MetaItem`s to internally use `TokenStream` representations (which will make `MetaItem` interactions much nicer for macro writers once the new macro system is in place).

r? @nrc

### CODE REVIEW

**Comment by eddyb** [comment]
cc @Manishearth

**Comment by Manishearth** [comment]
Shouldn't be breaking cc @dtolnay

File: src/librustc/lint/context.rs [hunk 1]
```diff
@@ -366,19 +366,19 @@ pub fn gather_attr(attr: &ast::Attribute)
...
     let meta = &attr.node.value;
-    let metas = match meta.node {
-        ast::MetaItemKind::List(_, ref metas) => metas,
-        _ => {
-            out.push(Err(meta.span));
-            return out;
-        }
-    };
+    let metas = if let Some(metas) = meta.meta_item_list() {
+                    metas
```
Comment (eddyb) [inline src/librustc/lint/context.rs]: I think should be indented just once.

File: src/librustc/lint/context.rs [hunk 2]
```diff
@@ -366,19 +366,19 @@ pub fn gather_attr(attr: &ast::Attribute)
...
+                    return out;
+                };
 
     for meta in metas {
-        out.push(match meta.node {
-            ast::MetaItemKind::Word(ref lint_name) => Ok((lint_name.clone(), level, meta.span)),
-            _ => Err(meta.span),
-        });
+        out.push(if meta.is_word() {
+                     Ok((meta.name().clone(), level, meta.span))
```
Comment (eddyb) [inline src/librustc/lint/context.rs]: Same (indent).

File: src/librustc_driver/lib.rs [hunk 3]
```diff
@@ -640,20 +638,19 @@ impl RustcDefaultCalls {
...
-                            ast::MetaItemKind::Word(ref word) => println!("{}", word),
-                            ast::MetaItemKind::NameValue(ref name, ref value) => {
-                                println!("{}=\"{}\"", name, match value.node {
-                                    ast::LitKind::Str(ref s, _) => s,
-                                    _ => continue,
-                                });
+                        if cfg.is_word() {
+                            println!("{}", cfg.name());
+                        } else if cfg.is_value_str() {
+                            let rhs = cfg.value_str();
```
Comment (eddyb) [inline src/librustc_driver/lib.rs]: Can't this be `} else if let Some(s) = cfg.value_str() {`?

File: src/librustc_incremental/assert_dep_graph.rs [hunk 4]
```diff
@@ -110,13 +110,13 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
...
-                                meta_item.span,
-                                &format!("unexpected meta-item {:?}", meta_item.node));
-                        }
+                    if meta_item.is_word() && id.is_none() {
+                        id = Some(meta_item.name().clone());
+                    } else {
+                        // FIXME better-encapsulate meta_item (don't directly access `node`)
+                        self.tcx.sess.span_err(
+                            meta_item.span(),
+                            &format!("unexpected meta-item {:?}", meta_item.node));
```
Comment (eddyb) [inline src/librustc_incremental/assert_dep_graph.rs]: Shouldn't this be `span_bug!`? cc @nikomatsakis

File: src/librustc_incremental/assert_dep_graph.rs [hunk 5]
```diff
@@ -127,16 +127,15 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
...
-                        }
+                    if meta_item.is_word() && dep_node_interned.is_none() {
+                        dep_node_interned = Some(meta_item.name().clone());
+                    } else if meta_item.is_word() && id.is_none() {
+                        id = Some(meta_item.name().clone());
+                    } else {
+                        // FIXME better-encapsulate meta_item (don't directly access `node`)
+                        self.tcx.sess.span_err(
+                            meta_item.span(),
+                            &format!("unexpected meta-item {:?}", meta_item.node));
```
Comment (eddyb) [inline src/librustc_incremental/assert_dep_graph.rs]: Same (`span_bug!`).

File: src/librustc_lint/builtin.rs [hunk 6]
```diff
@@ -299,9 +299,10 @@ impl MissingDoc {
         }
 
         let has_doc = attrs.iter().any(|a| {
-            match a.node.value.node {
-                ast::MetaItemKind::NameValue(ref name, _) if *name == "doc" => true,
-                _ => false
+            if a.is_value_str() && a.name() == "doc" {
+                true
```
Comment (eddyb) [inline src/librustc_lint/builtin.rs]: This `if`-`else` can be just the condition.

File: src/librustc_metadata/encoder.rs [hunk 7]
```diff
@@ -1431,31 +1431,28 @@ fn encode_item_index(rbml_w: &mut Encoder, index: IndexData) {
 }
 
 fn encode_meta_item(rbml_w: &mut Encoder, mi: &ast::MetaItem) {
```
Comment (eddyb) [inline src/librustc_metadata/encoder.rs]: I wonder why this doesn't use auto-derived serialization.

File: src/librustdoc/clean/mod.rs [hunk 8]
```diff
@@ -498,21 +498,20 @@ pub enum Attribute {
...
-                NameValue(s.to_string(), lit_to_string(v))
-            }
-        }
+        if self.is_word() {
+            Word(self.name().to_string())
+        } else if let Some(v) = self.value_str() {
+            NameValue(self.name().to_string(), v.to_string())
+        } else { // must be a list
+            let l = self.meta_item_list().unwrap();
+              List(self.name().to_string(), l.clean(cx))
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This line seems misaligned.

File: src/libsyntax/attr.rs [hunk 9]
```diff
@@ -92,6 +92,13 @@ pub trait AttrMetaMethods {
     /// Gets a list of inner meta items from a list MetaItem type.
     fn meta_item_list(&self) -> Option<&[P<MetaItem>]>;
 
+    /// Indicates if the attribute is a Word.
+    fn is_word(&self) -> bool;
+    /// Indicates if the attribute is a Value String.
+    fn is_value_str(&self) -> bool;
+    /// Indicates if the attribute is a Meta-Item List.
+    fn is_meta_item_list(&self) -> bool;
```
Comment (eddyb) [inline src/libsyntax/attr.rs]: Can't the latter 2 have default implementations? Or maybe it's not even desirable for them to exist when `value_str` and `meta_item_list` exist.

File: src/librustc_incremental/assert_dep_graph.rs (hunk 4 above)
Comment (nrc) [inline src/librustc_incremental/assert_dep_graph.rs]: Yes, should be, I think.

File: src/librustdoc/clean/mod.rs [hunk 10]
```diff
@@ -2568,25 +2589,25 @@ impl ToSource for syntax_pos::Span {
...
+//             res.push('\'');
+//             res
+//         },
+//         ast::LitKind::Char(c) => format!("'{}'", c),
+//         ast::LitKind::Int(i, _t) => i.to_string(),
+//         ast::LitKind::Float(ref f, _t) => f.to_string(),
+//         ast::LitKind::FloatUnsuffixed(ref f) => f.to_string(),
+//         ast::LitKind::Bool(b) => b.to_string(),
+//     }
+// }
```
Comment (nrc) [inline src/librustdoc/clean/mod.rs]: If we don't need this code then delete it.

File: src/libsyntax/attr.rs [hunk 11]
```diff
@@ -194,22 +222,38 @@ impl AttributeMethods for Attribute {
...
 }
 
 pub fn mk_word_item(name: InternedString) -> P<MetaItem> {
-    P(dummy_spanned(MetaItemKind::Word(name)))
+    mk_spanned_word_item(DUMMY_SP, name)
+}
+
+pub fn mk_spanned_name_value_item(sp: Span, name: InternedString, value: ast::Lit)
+                          -> P<MetaItem> {
+    P(respan(sp,MetaItemKind::NameValue(name, value)))
```
Comment (nrc) [inline src/libsyntax/attr.rs]: nit: space after comma (and below)

File: src/librustc_incremental/calculate_svh.rs [hunk 12]
```diff
@@ -69,7 +70,7 @@ impl<'a, 'tcx> SvhCalculate for TyCtxt<'a, 'tcx, 'tcx> {
         // to avoid hashing the AttrId
         for attr in &krate.attrs {
             debug!("krate attr {:?}", attr);
-            attr.node.value.hash(&mut state);
+            attr.meta().hash(&mut state);
```
Comment (nrc) [inline src/librustc_incremental/calculate_svh.rs]: I find the name `meta` to be not useful, maybe you should rename it? (Could be a follow-up PR).

Comment (cgswords) [inline src/librustc_incremental/calculate_svh.rs]: Do you have name suggestions? `value()`? `meta_item()` (which is dangerously close to `meta_item_list()`?

Comment (nrc) [inline src/librustc_incremental/calculate_svh.rs]: `inner()` or `meta_item()` seem good

File: src/librustc/lint/context.rs [hunk 13]
```diff
@@ -367,18 +367,18 @@ pub fn gather_attr(attr: &ast::Attribute)
...
-                    metas
-                } else {
-                    out.push(Err(meta.span));
-                    return out;
-                };
+      metas
+    } else {
+      out.push(Err(meta.span));
+      return out;
+    };
```
Comment (nrc) [inline src/librustc/lint/context.rs]: nit: 2-space indent

**Comment by nrc** [comment]
r = me with indenting nit and tests fixed

**Comment by alexcrichton** [comment]
@bors: retry

* #34595

**Comment by cgswords** [comment]
@alexcrichton Will you kick off bors again?

**Comment by alexcrichton** [comment]
er,

@bors: r=nrc 5553901146fa80c652abdc514b38360a0ae7418d

**Pull request merged**